[Feature] Adds a transport class for the Http Input Plugin

jloehel opened the pull request:
This commit adds a more generic class called Transport and a additional
class HttpTransport to create a client for the Http Input Plugin of
logstash: https://www.elastic.co/guide/en/logstash/current/plugins-inputs-http.html#plugins-inputs-http

Signed-off-by: Jürgen Löhel <juergen.loehel@inlyse.com>

eht16 commented on logstash_async/transport.py:
```diff
@@ -176,3 +316,173 @@ def send(self, events, use_logging=False):
+        :type use_logging: bool
+        """
+        self.__session = requests.Session()
+        for event in events:
```
Does the HTTP input support receiving multiple events at once?
This would speed up the transmission a lot. Currently the code would perform one HTTP request per event.

jloehel replied:
You are right, multiline events are possible. I will add this and add some chunk mechanism based on the post_max_size.

eht16 commented on logstash_async/transport.py:
```diff
@@ -3,15 +3,155 @@
+        :return: The current name of the host
+        :rtype: str
+        """
+        return self.__host
```
Is there a reason for using the double score name mangling everywhere?
So far I can't see it's necessary.

If not, I would prefer normal single underscores to mark instance attributes as "semiprivate".

jloehel replied:
I know that there is no real visibility for attributes in python. Anyway I try to keep the attributes "private" not just "protected" if I use properties. From the user perspective, it's more clear for me.

_"Protected"_
```
>>> from logstash_async.transport import HttpTransport
>>> test = HttpTransport('127.0.0.1', '5959')
>>> test.host
'127.0.0.1'
>>> test._host
'127.0.0.1'
>>> test._host = '127.0.0.2'
>>> test.host
'127.0.0.2'
```

_"Private"_
```
>>> from logstash_async.transport import HttpTransport
>>> test = HttpTransport('127.0.0.1', '5959')
>>> test.host
'127.0.0.1'
>>> test.__host
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'HttpTransport' object has no attribute '__host'
>>> test._Transport__host
'127.0.0.1'
>>> test._Transport__host = '127.0.0.2'
>>> test.host
'127.0.0.2'
```

eht16 replied:
Ok, fine by me.

eht16 commented on logstash_async/transport.py:
```diff
@@ -3,15 +3,155 @@
+        self.__host = name
+
+    @property
+    def port(self) -> int:
```
I would prefer to use properties when necessary, not always. It just clutters the code.
They can easily be added later if necessary.

jloehel replied:
I understand your point. I thought about raising exceptions for wrong inputs but let's keep it simple. I will convert all direct variables.

eht16 commented on logstash_async/transport.py:
```diff
@@ -3,15 +3,155 @@
+        self.codec = codec
+        self.ssl_enable = ssl_enable
+        self.ssl_verify = ssl_verify
+        super().__init__()
```
IMO all instance attributes should be declared in `__init__()`.

jloehel replied:
What attributes are missing?

jloehel commented:
> Thanks for contribution.
> I added a few remarks and questions, it would be nice to get them resolved.

Thank you very much for your review. I will answer your question and I will add the documentation as quickly as possible.

jloehel commented on logstash_async/transport.py:
```diff
@@ -3,15 +3,155 @@
+        self.__timeout = time
+
+    @property
+    def codec(self) -> str:
```
You are right. I will remove the codec attribute and document that the HttpInput Transport sends the logs always in json.

jloehel commented on logstash_async/transport.py:
```diff
@@ -3,15 +3,155 @@
 
+from abc import ABC, abstractmethod
+from time import sleep
+from typing import Union
```
In my opinion the break with python2 needs to be [strict](https://python3statement.org) now. The people had enough time (5 years) to prepare for the change. But if you really want to keep the python2 support, I will revert the changes.

eht16 replied:
While I agree that Python2 should not be used anymore and the transition time was long enough, I also see the reality: even with this long transition, still not everyone had to chance to switch to Python3.
The main issue with dropping Python2 support is that it will hit the developers who already had no time to migrate, it will not hit the people who made this decision.
I know at least one company using this library affected.

Obviously, keeping Python2 support won't work in the long term and it doesn't make sense to do so.
But I want to keep it at least for some more time. But then finally drop Python2 support and make a major version bump.

eht16 replied:
+ thanks for your efforts on reverting the Python3 only changes.

jloehel commented:
@eht16 I changed the code. Can you please review it again.

jloehel commented:
They release isort 5.0.4 yesterday. [They removed SortImports](https://github.com/timothycrosley/isort/commit/c7378c3b77a115801200e7a20201075858997a28).  But pylint is still using it.

eht16 commented:
> They released isort 5.0.4 yesterday and [ removed SortImports](https://github.com/timothycrosley/isort/commit/c7378c3b77a115801200e7a20201075858997a28). But pylint is still using it.

I know. There is https://github.com/PyCQA/pylint/issues/3722 and https://github.com/PyCQA/pylint/pull/3725.
I hope a compatible Pylint version will be released soon.

eht16 commented:
> edit: Usually I use now the black code style. But I will convert all the double quotes back into single quotes.

That's one of the reasons I don't use `black`: they force people to use their style. I got the main idea of establishing a common code style and that they think double quotes are better or whatever but I'd like to be able to choose. More generally, the lack of configuration options by design in `black` makes it unusable for me.
So I tend to stick with `isort`, `flake8` and `pylint`.

eht16 commented on logstash_async/transport.py:
```diff
@@ -176,3 +224,117 @@ def send(self, events, use_logging=False):
+            self,
+            host,
+            port,
+            timeout=2.0,
```
We could re-use the timeout logic from the other transports here: take `TimeoutNotSet` as default and set the tmeout only when a real timeout is passed. When the transport is setup by the handler, a default timeout of five seconds (as defined in `constants.SOCKET_TIMEOUT`) is set.

jloehel replied:
Sure, I will do it tomorrow.

eht16 commented on logstash_async/transport.py:
```diff
@@ -176,3 +224,117 @@ def send(self, events, use_logging=False):
+            protocol = 'https'
+        return '{}://{}:{}'.format(protocol, self.host, self.port)
+
+    def encode(self, events):
```
Could be private. But doesn't make much difference, so skip if you want.

jloehel replied:
You are right, it should be private.

eht16 commented on logstash_async/transport.py:
```diff
@@ -3,15 +3,63 @@
+            ssl_enable=True,
+            ssl_verify=True
+    ):
+        self.host = host
```
Why not making the instance attributes private or in this maybe better semiprivate?
Depending on your decision, also for `HttpTransport.username` and `HttpTransport.password`.

jloehel replied:
Of course the attributes should be private but then I would need public functions to access the variables. Without properties or getters/setters I would prefer protected.

eht16 replied:
When I say "semiprivate" I mean what you mean by "protected", just to be sure we are talking about the same.

Do you need to access the instance attributes from outside? If I didn't miss anything, they are accessed only in the base and sub class? So semiprivate/protected should be fine.

eht16 commented on logstash_async/transport.py:
```diff
@@ -176,3 +224,117 @@ def send(self, events, use_logging=False):
+                attempt += 1
+            else:
+                self.close()
+                error_msg = 'Logstash respond with error {}'.format(status_code)
```
`responded` might be better as we got the response already?
More importantly, we should also include the text response from Logstash. I assume this will help more than just reading "there was a 500" :).

jloehel replied:
I will check how the plugin handles errors in general and try to add the message here.

eht16 replied:
Maybe it's enough to simply add `requests.text`.

jloehel replied:
I use now `response.reason`

eht16 commented on logstash_async/transport.py:
```diff
@@ -176,3 +224,117 @@ def send(self, events, use_logging=False):
+            status_code = response.status_code
+            if status_code == 200:
+                break
+            if status_code == 429:
```
Basically, we don't need to handle 429 and backoff explicitly.
The caller (`LogProcessingWorker`) will by design handle errors which are raised here and requeue all events passed to `send()`. The other transports also don't have own error handling as it is done by the caller more gracefully.

jloehel replied:
What exception should be raised in case of an error?

eht16 replied:
It doesn't matter much which exception is thrown. See https://github.com/eht16/python-logstash-async/blob/master/logstash_async/worker.py#L212, `Exception` is catched.
Raising a `RuntimeError` as you did should be fine.

The idea of the exception handling in the worker (as mentioned above) is that the worker will detect if there was an error on delivery via the transport and in this case, the events are put back to the database so that delivery will be tried again. By putting the events back to the database, they are also persisted in case the application shuts down between the failed delivery attempt and the next attempt.
While there is no real backoff logic, the `QUEUED_EVENTS_FLUSH_` settings in `constants.py` can be used to tune the sending behavior.
And this works for all transports in the same way.

eht16 replied:
It might be even easier to just use `response.raise_for_status()`, this will raise a proper exception for any non-success status. The exceptions look like:
```
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://example.com/non-existant
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://example.com/admin/
```
So not message composition and status code handling is necessary.
Just an idea.

jloehel replied:
I like the idea. :-)

jloehel commented:
> > edit: Usually I use now the black code style. But I will convert all the double quotes back into single quotes.
> 
> That's one of the reasons I don't use `black`: they force people to use their style. I got the main idea of establishing a common code style and that they think double quotes are better or whatever but I'd like to be able to choose. More generally, the lack of configuration options by design in `black` makes it unusable for me.
> So I tend to stick with `isort`, `flake8` and `pylint`.

thanks for the explaination

jloehel commented:
Sorry, I closed it accidentally ^^

eht16 commented:
> > They released isort 5.0.4 yesterday and [ removed SortImports](https://github.com/timothycrosley/isort/commit/c7378c3b77a115801200e7a20201075858997a28). But pylint is still using it.
> 
> I know. There is [PyCQA/pylint#3722](https://github.com/PyCQA/pylint/issues/3722) and [PyCQA/pylint#3725](https://github.com/PyCQA/pylint/pull/3725).
> I hope a compatible Pylint version will be released soon.

If you want, you can apply the following changes to tox.ini (and commit them) as a temporary workaround to get the checks running again until Pylint is fixed:
```diff
diff --git a/tox.ini b/tox.ini
index 9f1586c..62995f6 100644
--- a/tox.ini
+++ b/tox.ini
@@ -13,7 +13,7 @@ logstash_async_modules = logstash_async tests
 [testenv]
 deps =
     flake8
-    isort
+    isort<5
     pylint
 commands =
     # linting and code analysis
@@ -29,7 +29,7 @@ commands =
 # for Python 2.7.
 deps =
     flake8
-    isort
+    isort<5
 commands =
     # linting and code analysis
     {envbindir}/flake8 {[tox]logstash_async_modules}
```

jloehel commented:
@eht16 I added the documentation for HttpTransport and I implemented the requested changes. I will add some tests during this week or maybe next week.

eht16 commented on example3.py:
```diff
@@ -0,0 +1,43 @@
+date_format = '%Y-%m-%d | %H:%M:%S'
+formatter = logging.Formatter(logging_format, date_format)
+handler.setFormatter(formatter)
+logger.addHandler(handler)
```
`logger` and so also `handler` and `formatter` are not used at all, so can be removed.

jloehel replied:
IThis is the configuration for the logstash_async.transport logger. To see the debugging logs from there.

eht16 replied:
Ah, sorry.

eht16 commented on logstash_async/transport.py:
```diff
@@ -176,3 +227,144 @@ def send(self, events, use_logging=False):
+            if len(current_event) > self._max_content_length:
+                msg = 'The event size <%s> is greater than the max content length <%s>.'
+                msg += 'Skipping event.'
+                logger.warning(msg, len(current_event), self._max_content_length)
```
Logging in transports can be tricky: if `logging.shutdown()` has been called, the logging framework basically doesn't work anymore and we have to just finish transmitting.
This is why the handler passes the `use_logging` argument to tell the transport whether it might use the logging framework or not.
See https://github.com/eht16/pylogbeat/blob/master/pylogbeat.py#L72 for some inspiration.

jloehel replied:
Does the people use the AsynchronousLogstashHandler for the root logger or using the handler for the logstash_async itself? But you are right, it can be tricky. But I would not like to skip the event silently. Should we let it fail and just do not care about the corner case? The bad thing is that all future events will fail, too. Right?

eht16 replied:
The idea is to use `AsynchronousLogstashHandler` in applications as a handler for the existing logging configuration and I guess usually for the root logger. So this handler primarily handles the application's log messages.

I agree to not silently drop events on errors. There is nothing wrong with this warning message but I wanted to suggest to issue it only if `use_logging` is set and to print to `stderr` otherwise.

It might be good to have `_safe_log` (https://github.com/eht16/python-logstash-async/blob/f09055373613c988f6ab4a9e485bb8b12598cafd/logstash_async/worker.py#L267) also available in the transports for this. But I can work on this after this PR, we don't need to refactor the world all in here :D.

eht16 commented on logstash_async/transport.py:
```diff
@@ -176,3 +227,144 @@ def send(self, events, use_logging=False):
+            protocol = 'https'
+        return '{}://{}:{}'.format(protocol, self._host, self._port)
+
+    def __batches__(self, events):
```
Sorry for being a pit picky: should be `__batches`, IMO. The trailing double underscore would make it a special method (https://docs.python.org/3/reference/datamodel.html#special-method-names) which is rather confusing.

jloehel replied:
Sure, I change it

jloehel commented on logstash_async/transport.py:
```diff
@@ -262,6 +267,7 @@ def __init__(
         super().__init__(host, port, timeout, ssl_enable, ssl_verify)
         self._username = kwargs.get('username', None)
         self._password = kwargs.get('password', None)
+        self._max_content_length = kwargs.get('max_content_length', 104857600)
```
Sure, I change it

jloehel commented:
> Thanks again for your efforts!
> Especially for the great docs and consistently using "TLS" 100

You are welcome.

eht16 commented on logstash_async/transport.py:
```diff
@@ -176,3 +227,144 @@ def send(self, events, use_logging=False):
+        self.__session = requests.Session()
+        for batch in self.__batches(events):
+            logger.debug('Batch length: %s', len(batch))
+            logger.debug('Batch size: %s', len(json.dumps(batch).encode('utf8')))
```
These two messages could be combined to one.

jloehel commented:
@eht16 Sorry, I was busy last week. I added the the requested changes. I hope I'll find some time to implement the tests this week.

eht16 commented:
Hi @jloehel,

any news?
Fortunately, the isort vs pylint problem has been resolved in the meantime, so the changes in `tox.ini` are no longer necessary.

jloehel commented:
> Hi @jloehel,
> 
> any news?
> Fortunately, the isort vs pylint problem has been resolved in the meantime, so the changes in `tox.ini` are no longer necessary.

Moin moin,

sorry I was some kind of busy the last weeks. I rebased the branch and reverted the isort/py27 commits. May I revert also the type hints and the f-strings?

eht16 commented:
> > Hi @jloehel,
> > any news?
> > Fortunately, the isort vs pylint problem has been resolved in the meantime, so the changes in `tox.ini` are no longer necessary.
> 
> Moin moin,
> 
> sorry I was some kind of busy the last weeks. I rebased the branch and reverted the isort/py27 commits. May I revert also the type hints and the f-strings?

Yes, fine by me, sorry for the trouble. But at least now we are Python3 only.

jloehel commented:
> > > Hi @jloehel,
> > > any news?
> > > Fortunately, the isort vs pylint problem has been resolved in the meantime, so the changes in `tox.ini` are no longer necessary.
> > 
> > 
> > Moin moin,
> > sorry I was some kind of busy the last weeks. I rebased the branch and reverted the isort/py27 commits. May I revert also the type hints and the f-strings?
> 
> Yes, fine by me, sorry for the trouble. But at least now we are Python3 only.

Null problemo. ^^ I added the two commits. :-)

eht16 commented:
Thanks a lot for this feature and your patience with my remarks :).
I squashed most of the commits together to have it a little cleaner in the history and finally merged. Unfortunately, because of the manual squashing, Github doesn't consider this branch as merged but it is.

If you still want to add some tests for the HTTP transport, that'd be fine.
